### ISSUE TITLE
Double backwards 3 times slower for conv2d with padding = 1

### ISSUE THREAD

**Issue by MichaelArbel** [pytorch/pytorch#22577]
## 🐛 Bug

Hi, I'm trying to compute the jacobian vector product using a trick from https://j-towns.github.io/2017/06/12/A-new-trick.html. This involves computing the gradient twice. It seems like using conv2d with padding =1 slows down the computation by almost 3 times, compared to padding =0:

padding= 0:  2.962866 s
padding=1:  9.922703s

In case only one backward pass is computed both give the same times:

padding =0: 1.056120s
padding = 1: 1.180639s

## To Reproduce
import  torch as tr
from torch.autograd import Variable
import torch.nn as nn
from torch import autograd

import time
from copy import deepcopy
from datetime import datetime
startTime = datetime.now()

import torch.nn as nn
import torch.nn.functional as F
from models.resnet import *

tr.backends.cudnn.benchmark=True

class Net(nn.Module):
    def __init__(self,padding= 0):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(3, 64, 3,padding=padding)
        self.conv2 = nn.Conv2d(64, 64, 3,padding=padding)

    def forward(self, x):
        out = F.relu(self.conv1(x))
        out = F.relu(self.conv2(out))
        return out.sum()

def jvp(outputs,variable,direction):
    dummy_u =  Variable(tr.ones_like(outputs), requires_grad=True)
    aux_grad_loss = tr.autograd.grad(outputs = outputs,inputs = variable, grad_outputs=dummy_u,  create_graph=True,only_inputs=True)
    out = tr.autograd.grad(outputs= aux_grad_loss,inputs = dummy_u,grad_outputs=direction,create_graph=True,only_inputs=True)[0]
    return out

device = 'cuda'
dist = tr.distributions.Normal(tr.tensor(4.0,dtype= tr.float32, device = device), tr.tensor(0.5,dtype= tr.float32, device = device))

b_size = 128
num_iter = 100
X = dist.sample([b_size,3,32,32])

print("no padding")
net = Net(padding=0).to(device)

params = list(net.parameters())
directions = deepcopy(params)

torch.cuda.synchronize()
startTime = datetime.now()
for k in range(num_iter):

    Y = tr.sum(net(X))
    dd = jvp(Y,params,directions)

torch.cuda.synchronize()
print(datetime.now() - startTime)

print("with padding")

net = Net(padding=1).to(device)
params = list(net.parameters())
directions = deepcopy(params)

torch.cuda.synchronize()
startTime = datetime.now()
for k in range(num_iter):
    Y = tr.sum(net(X))
    dd = jvp(Y,params,directions)
torch.cuda.synchronize()
print(datetime.now() - startTime)

## Expected behavior

The computation time should be roughly similar with or without padding.

## Environment

PyTorch version: 1.1.0
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 16.04.5 LTS
GCC version: (crosstool-NG 1.23.0.449-a04d0) 7.3.0
CMake version: version 3.7.2

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: Quadro P5000
GPU 1: GeForce RTX 2080

Nvidia driver version: 410.73
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.3.1

Versions of relevant libraries:
[pip] numpy==1.16.4
[pip] torch==1.1.0
[pip] torchvision==0.3.0
[conda] mkl                       2019.4                      243
[conda] torch                     1.1.0                     <pip>
[conda] torchvision               0.3.0                     <pip>

**Comment by ezyang** [comment]
@fmassa: This might just be cuDNN dispatching to a bad kernel in the padding = 1. Can you run your code under nvprof and let us know which kernels it is running in this case?

**Comment by MichaelArbel** [comment]
Hi @ezyang ,
Thanks for your quick reply.
So when I run nvprof with torch.backends.cudnn.benchmark=False, the difference is even larger:

- without padding: 4.70s
![Screen Shot 2019-07-19 at 00 52 40](https://user-images.githubusercontent.com/12251667/61499630-d346f000-a9bf-11e9-992f-f5e07d27a087.png)

- with padding: 1 min 20.76s
![Screen Shot 2019-07-19 at 00 51 02](https://user-images.githubusercontent.com/12251667/61499653-eeb1fb00-a9bf-11e9-93a8-fd9eb2921610.png)

Now with  torch.backends.cudnn.benchmark=True, I get:

- without padding: 2.59s
![Screen Shot 2019-07-19 at 00 59 25](https://user-images.githubusercontent.com/12251667/61499870-a6dfa380-a9c0-11e9-8496-25a171f802be.png)

- with padding: 4.40s:

![Screen Shot 2019-07-19 at 01 00 22](https://user-images.githubusercontent.com/12251667/61499887-bd85fa80-a9c0-11e9-9d16-969ae1850771.png)


This is better than last time for which  it took 9.92s even in benchmark mode, it might be because the Nvidia driver was updated meanwhile: 
Nvidia driver version: 410.78

**Comment by fps7806** [comment]
I'm running into the same issue as Michael with the following code:
```
import time
import torch
from torch import nn
from torch.autograd import grad as grad_f


test = torch.zeros(128, 32, 32, 32, device='cuda')

n = 100
conv = nn.Conv2d(32, 128, kernel_size=3, padding=1).cuda()
torch.cuda.synchronize()
tstart = time.time()
for _ in range(n):
    x = conv(test)
torch.cuda.synchronize()
print("forward time:", time.time() - tstart)

grad = torch.zeros_like(x)
torch.cuda.synchronize()
tstart = time.time()
for _ in range(n):
    x = conv(test)
    dw, = grad_f(x, conv.weight, grad_outputs=grad, create_graph=True, allow_unused=False)
torch.cuda.synchronize()
print("forward + backward time:", time.time() - tstart)

second_grad = torch.zeros_like(conv.weight)
torch.cuda.synchronize()
tstart = time.time()
for _ in range(n):
    x = conv(test)
    dw, = grad_f(x, conv.weight, grad_outputs=grad, create_graph=True, allow_unused=False)
    dw.backward(second_grad)
torch.cuda.synchronize()
print("forward + backward * 2 time:", time.time() - tstart)
```

The slowdown in my case is closer to 30x of what I would expect. And it seems like the 2nd order gradient is computed with some weird kernels, this is what I get from using nvprof:
```
Profiling result:
            Type  Time(%)      Time     Calls       Avg       Min       Max  Name
 GPU activities:   58.48%  6.95173s     30720  226.29us  197.28us  261.60us  volta_gcgemm_32x32_nt
                   16.04%  1.90673s     28800  66.206us  42.336us  96.031us  void fft2d_r2c_32x32<float, bool=0, unsigned int=1, bool=0>(float2*, float const *, int, int, int, int, int, int, int, int, int, cudnn::reduced_divisor, bool, int2, int, int)
                   16.02%  1.90446s     30720  61.994us  41.919us  87.455us  void fft2d_c2r_32x32<float, bool=0, bool=0, unsigned int=0, bool=0, bool=0>(float*, float2 const *, int, int, int, int, int, int, int, int, int, float, float, cudnn::reduced_divisor, bool, floa
t*, float*, int2, int, int)
                    4.42%  525.74ms        60  8.7624ms  8.7531ms  8.7715ms  void cudnn::detail::implicit_convolve_sgemm<float, float, int=1024, int=5, int=5, int=3, int=3, int=3, int=1, bool=1, bool=0, bool=1>(int, int, int, float const *, int, float*, cudnn::detail::i
mplicit_convolve_sgemm<float, float, int=1024, int=5, int=5, int=3, int=3, int=3, int=1, bool=1, bool=0, bool=1>*, kernel_conv_params, int, float, float, int, float, float, int, int)
                    1.15%  136.26ms      1950  69.875us  44.640us  321.37us  void fft2d_r2c_32x32<float, bool=0, unsigned int=0, bool=0>(float2*, float const *, int, int, int, int, int, int, int, int, int, cudnn::reduced_divisor, bool, int2, int, int)
```

Interestingly enough, if I take my code and use `.to(torch.float64)` the problem goes away, so it seems like only the float32 version has issues.

I can't identify where the 2nd order gradients would be defined as I am not familiar with pytorch's source code, but I'm happy to take a look if I get some pointers.

**Comment by fps7806** [comment]
Note: the problem goes away if I change the shape of the input or stride of the convolution

**Comment by fps7806** [comment]
Sorry for spamming the thread, but I found out that setting `torch.backends.cudnn.deterministic = True` significantly improved the performance of my script. @MichaelArbel can you try it in your script?

**Comment by jotaf98** [comment]
I have the same problem, even worse (120x slowdown):

```
forward time: 0.6090080738067627
forward + backward time: 0.49439048767089844
forward + backward * 2 time: 61.386539697647095
```

Using float64 somewhat improves the *relative* slowdown, but of course is very slow in absolute terms -- so I would say the "has workaround" label is a bit optimistic...

```
forward time: 4.89356541633606
forward + backward time: 9.84079647064209
forward + backward * 2 time: 30.34150195121765
```

**Comment by ezyang** [comment]
@jotaf98 did turning on deterministic help you rcase?

**Comment by kurtamohler** [comment]
If anyone else is working on this issue, please let me know. Otherwise, I'm starting to look into it.

**Comment by ngimel** [comment]
There's not much to work on here, setting torch.backends.cudnn.benchmark to True usually makes the problem much less severe, and beyond that not much can be done, because convolutions are thin wrappers around cudnn convolutions.

**Comment by kurtamohler** [comment]
@ngimel, even using `torch.backends.cudnn.benchmark = True` and `torch.backends.cudnn.deterministic = True`, the second gradient calculation takes 3x time than the first gradient calculation. Even if it makes the problem less severe, it's still a pretty significant difference.

Furthermore, slowdown shows up in the CPU implementation as well:
```
forward time: 3.184596538543701
forward + backward time: 6.761305570602417
forward + backward * 2 time: 40.367207527160645
second backward slowdown: 9.395761x
```

That seems like evidence that this issue is not inside cudnn--it's in pytorch itself. What are your thoughts?

**Comment by ngimel** [comment]
Double backward for convolution just calls convolutions, just when all the gradients are required there are more convolutions, but those computations are necessary. 
https://github.com/pytorch/pytorch/blob/f531815526c69f432e46fadece44f5d3a9b70e30/aten/src/ATen/native/Convolution.cpp#L763-L998.
Before digging deep you should make sure that the results you are getting are really much worse than what you'd expect given the amount of compute double backward involves.  
For some cases with certain padding and strides, torch slow implementation of convolution is called instead of cudnn, I wonder if this benchmark is hitting one of those.

**Comment by kurtamohler** [comment]
@ngimel, ah I see what you mean, those extra convolution calculations probably do explain why double backward is taking much longer than the single backward. Thanks for the information!

> For some cases with certain padding and strides, torch slow implementation of convolution is called instead of cudnn, I wonder if this benchmark is hitting one of those.

If that is true, do you think it's necessary to call the torch implementation, or would it be possible to change it to use cudnn?

**Comment by ngimel** [comment]
Currently if `is_output_padding_big` returns true cudnn is not used. https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Convolution.cpp#L92-L98. At the time this was added cudnn did not handle those cases correctly, @ptrblck and @jjsjann123 would be the best people to check with cudnn if now it can handle those cases.

**Comment by ezyang** [comment]
Based on @ngimel's diagnosis I'm downgrading the priority on this issue
